File: Assets/Script/Loader/LoaderAssets/AssetBundleLoader.cs
using System;
using System.Collections;
using UnityEngine;
using UnityEngine.SceneManagement;

namespace Hukiry
{
    /// <summary>
    /// 资源加载管理
    /// </summary>
    public class AssetBundleLoader : MonoBehaviour, IBaseLoader {

		/// <summary>
		/// 当前的实体
		/// </summary>
		internal static LoaderAsyncAsset Instance { get; private set; }
        #region 资源依赖相关

        /// <summary>
        /// 获取AssetBundle的依赖，会缓存
        /// </summary>
        /// <param name="name"></param>
        /// <returns></returns>
        internal static string[] GetDependencies(string name) {
			HotAssetInfo mData = GetManifestData(name);
			if (mData != null) {
				return mData.ds.ToArray();
			}
			return new string[0] { };
		}

		/// <summary>
		/// 获取AssetBundle的依赖，会缓存
		/// </summary>
		/// <param name="name"></param>
		/// <returns></returns>
		public static HotAssetInfo GetManifestData(string abName) {
			return AssetBundleList.ins.GetAssetBundleDataAtPath(abName);
		}

        /// <summary>
        /// 查找路径
        /// </summary>
        /// <param name="resName"></param>
        /// <returns></returns>
        public static string FindPath(string resName) {
			HotAssetInfo mData = GetManifestData(resName);
			if (mData != null) {
                return mData.GetPath();

            }
			return resName;
		}

		#endregion

		/// <summary>
		/// 创建加载器
		/// </summary>
		/// <param name="parent"></param>
		/// <returns></returns>
		public static AssetBundleLoader Create(Transform parent)
        {
            GameObject go = new GameObject("AssetBundleLoader");
            
            Instance = go.AddComponent<LoaderAsyncAsset>();

            DontDestroyOnLoad(go);

            return go.AddComponent<AssetBundleLoader>();
        }

        /// <summary>
        /// 加载资源
        /// </summary>
        /// <param name="assetName">资源路径</param>
        /// <param name="abName">ab路径</param>
        /// <returns>资源对象</returns>
        public UnityEngine.Object LoadAsset(string assetName, string abName)
        {
            return LoaderLocalAsset.LoadAssets(assetName, abName);
        }

        /// <summary>
        /// 同步加载列表资源
        /// </summary>
        /// <param name="assetName">资源路径</param>
        /// <param name="abName">ab路径</param>
        /// <returns></returns>
        public object[] LoadAllAsset(string assetName, string abName)
        {
            return LoaderLocalAsset.LoadAssetsList(assetName, abName);
        }

        /// <summary>
        /// 异步加载资源
        /// </summary>
        /// <param name="assetName">资源路径</param>
        /// <param name="abName">ab路径</param>
        /// <param name="OnLoadedCallback">加载完成回调</param>
        public void LoadAsync(string assetName, string abName, OnAssetLoadedCallback<UnityEngine.Object> OnLoadedCallback, bool loadObject)
        {
            AssetBundleList.ins.Log("LoadAsync",assetName, abName);
            //启动异步加载
            Instance.LoadAssets(assetName, abName, OnLoadedCallback, loadObject);
        }

        /// <summary>
        /// 异步加载资源列表
        /// </summary>
        /// <param name="assetName">资源路径</param>
        /// <param name="abName">ab路径</param>
        /// <param name="OnLoadedCallback">加载完成回调</param>
        public void LoadAllAsync(string assetName, string abName, OnAssetListLoadedCallback<UnityEngine.Object> OnLoadedCallback)
        {
            //启动异步加载
            Instance.LoadAssets<UnityEngine.Object>(assetName, abName, OnLoadedCallback);
        }

        /// <summary>
        /// 异步加载场景
        /// </summary>
        /// <param name="levelName"></param>
        /// <param name="onProgress"></param>
        public void LoadSceneAsync(string levelName, string abName, Action<float> onProgress, Action onFinish)
        {
            StartCoroutine(_LoadSceneAsync(levelName, abName, onProgress, onFinish));
        }

        IEnumerator _LoadSceneAsync(string levelName, string abName, Action<float> onProgress, Action onFinish)
        {
            float proportion = 0.5f;
            float displayProgress = 0;

			if (AssetBundleList.ins.GetAssetBundleDataAtPath(abName) != null) {
				//先加载ab文件
				bool abLoadFinish = false;
				LoadAsync(levelName, abName, (name, obj) => {
					abLoadFinish = true;
				}, true);

				while (!abLoadFinish) {
                    displayProgress = Mathf.Min(displayProgress + 0.01f, proportion);
                    onProgress(displayProgress);
					yield return new WaitForEndOfFrame();
				}
			}

            //从内存加载
            AsyncOperation op = SceneManager.LoadSceneAsync(levelName);

            while (!op.isDone) {
                onProgress(op.progress * proportion + proportion);
                yield return new WaitForEndOfFrame();
            }
			AssetsLoaderMgr.Unload(abName, 0, true);

            onFinish?.Invoke();
		}

        /// <summary>
        /// 加载失败,释放已经加载的资源
        /// </summary>
        /// <param name="path"></param>
        /// <param name="parent"></param>
        internal static void FailReleaseLoaded(string path, string parent)
        {
            //遍历子节点
            string[] dependices = GetDependencies(path);
            for (int i = 0; i < dependices.Length; ++i)
            {
                FailReleaseLoaded(dependices[i], path);
            }

            //如果存在释放parent的引用
            AssetBundleObject obj = AssetBundleCache.FindAssetBundle(path);
            if (obj != null)
            {
                obj.DelReference(parent);
            }
        }
        /// <summary>
        /// 卸载未使用的资源
        /// </summary>
        public void UnloadUnusedAssets(bool gc)
        {
            AssetBundleCache.UnloadUnuseAsset(gc);
        }

        /// <summary>
        /// 清理所有资源
        /// </summary>
        public void UnLoadAllAssets()
        {
            AssetBundleCache.UnloadAllAsset();
        }

        void OnApplicationQuit()
        {
            this.UnLoadAllAssets();
        }
	}
}

File: LuaScript/Library/Loader/Res/Weg/AssetBundleLoader.lua
---
--- AssetBundleLoader       
--- Author : hukiry     
--- DateTime 2024/8/8 14:25   
---

---@class AssetBundleLoader
local AssetBundleLoader = Class()
---初始化字段
function AssetBundleLoader:ctor()

end

---加载资源
function AssetBundleLoader:LoadAsset(assetName,abName)
    return SingleAssist.AssetBlock():LoadAssets(assetName, abName)
end

---同步加载列表资源
---@return UnityEngine.Object[]
function AssetBundleLoader:LoadAllAsset(assetName,abName)
    return SingleAssist.AssetBlock():LoadAssetsList(assetName, abName)
end

---异步加载资源
function AssetBundleLoader:LoadAsync(assetName,abName,OnLoadedCallback,loadObject)
    SingleAssist.AssetAsync():LoadAssets(assetName, abName, OnLoadedCallback, loadObject)
end

---异步加载资源列表
function AssetBundleLoader:LoadAllAsync(assetName,abName,OnLoadedCallback)
    SingleAssist.AssetAsync():LoadAllAsync(assetName, abName, OnLoadedCallback)
end

---异步加载场景
---@param onProgress function<number>
---@param onFinish function
function AssetBundleLoader:LoadSceneAsync( levelName,  abName,  onProgress,  onFinish)
    StartCoroutine(Handle(self, self._LoadSceneAsync, levelName, abName, onProgress, onFinish))
end

---@private
function AssetBundleLoader:_LoadSceneAsync(levelName, abName, onProgress, onFinish)
    local proportion, displayProgress = 0.5, 0
    if SingleAssist.AssetBundle():GetAssetBundleDataAtPath(abName) then
        local abLoadFinish = false
        self:LoadAsync(levelName, abName, function(name, obj)
            abLoadFinish = true
        end, true)

        while(not abLoadFinish) do
            displayProgress = math.min(displayProgress + 0.01, proportion)
            onProgress(displayProgress)
            WaitForEndOfFrame()
        end
    end

    local op  = UnityEngine.SceneManagement.SceneManager.LoadSceneAsync(levelName)
    while(not op.isDone) do
        onProgress(op.progress * proportion + proportion)
        WaitForEndOfFrame()
    end

    SingleAssist.LoaderMgr():Unload(abName, 0, true)
    if onFinish then
        onFinish()
    end
end

function AssetBundleLoader:Unload(assetName, delay) end
---立即卸载资源
function AssetBundleLoader:UnloadImmediate(assetName, force)
    AssetBundleCache:UnloadImmediate(assetName .. ".ab")
end
---卸载未使用的资源
function AssetBundleLoader:UnloadUnusedAssets(gc)
    AssetBundleCache:UnloadUnuseAsset(gc)
end
---清理所有资源
function AssetBundleLoader:UnLoadAllAssets()
    AssetBundleCache:UnloadAllAsset()
end

function AssetBundleLoader:Destroy() end

return AssetBundleLoader